pkg/cfn/builder/iam.go

```diff
@@ -182,172 +182,141 @@ func (n *NodeGroupResourceSet) addResourcesForIAM() error {
-			rs.spec.Status = &api.ClusterIAMServiceAccountStatus{
+func NewIAMRoleResourceSetForServiceAccount(spec *api.ClusterIAMServiceAccount, oidc *iamoidc.OpenIDConnectManager) *IAMRoleResourceSet {
+	return &IAMRoleResourceSet{
+		template:            cft.NewTemplate(),
+		attachPolicy:        spec.AttachPolicy,
+		attachPolicyARNs:    spec.AttachPolicyARNs,
+		serviceAccount:      spec.Name,
+		namespace:           spec.Namespace,
+		wellKnownPolicies:   spec.WellKnownPolicies,
+		roleName:            spec.RoleName,
+		permissionsBoundary: spec.PermissionsBoundary,
+		description: fmt.Sprintf(
+			"IAM role for serviceaccount %q %s",
+			spec.NameString(),
+			templateDescriptionSuffix,
+		),
+		oidc: oidc,
+		roleNameCollector: func(v string) error {
+			spec.Status = &api.ClusterIAMServiceAccountStatus{
 				RoleARN: &v,
 			}
 			return nil
 		},
-	})
-
-	if len(rs.spec.AttachPolicy) != 0 {
-		rs.template.AttachPolicy("Policy1", roleRef, rs.spec.AttachPolicy)
 	}
-
-	return nil
-}
-
-// RenderJSON will render iamserviceaccount stack as JSON
-func (rs *IAMServiceAccountResourceSet) RenderJSON() ([]byte, error) {
-	return rs.template.RenderJSON()
-}
-
-// GetAllOutputs will get all outputs from iamserviceaccount stack
-func (rs *IAMServiceAccountResourceSet) GetAllOutputs(stack cfn.Stack) error {
-	return rs.outputs.MustCollect(stack)
 }
 
 // IAMRoleResourceSet holds IAM Role stack build-time information
 type IAMRoleResourceSet struct {
-	template         *cft.Template
-	oidc             *iamoidc.OpenIDConnectManager
-	outputs          *outputs.CollectorSet
-	name             string
-	attachPolicyARNs []string
-	attachPolicy     api.InlineDocument
-	OutputRole       string
-	serviceAccount   string
-	namespace        string
+	template            *cft.Template
+	oidc                *iamoidc.OpenIDConnectManager
+	outputs             *outputs.CollectorSet
+	roleName            string
+	wellKnownPolicies   api.WellKnownPolicies
+	attachPolicyARNs    []string
+	attachPolicy        api.InlineDocument
+	roleNameCollector   func(string) error
+	OutputRole          string
+	serviceAccount      string
+	namespace           string
+	permissionsBoundary string
+	description         string
 }
 
-// NewIAMServiceAccountResourceSet builds IAM Role stack from the give spec
-func NewIAMRoleResourceSetWithAttachPolicyARNs(name, namespace, serviceAccount string, attachPolicyARNs []string, oidc *iamoidc.OpenIDConnectManager) *IAMRoleResourceSet {
-	return &IAMRoleResourceSet{
-		template:         cft.NewTemplate(),
-		attachPolicyARNs: attachPolicyARNs,
-		name:             name,
-		oidc:             oidc,
-		serviceAccount:   serviceAccount,
-		namespace:        namespace,
+// NewIAMRoleResourceSetForServiceAccount builds IAM Role stack from the give spec
+func NewIAMRoleResourceSetWithAttachPolicyARNs(name, namespace, serviceAccount, permissionsBoundary string, attachPolicyARNs []string, oidc *iamoidc.OpenIDConnectManager) *IAMRoleResourceSet {
```

Comment:
This seems to duplicate a lot from `WithAttachPolicy`, can they be cleaned up?

Reply:
:+1: